File: src/accounts.py
"""Account configuration — parse accounts.toml with provider presets."""

import argparse
import os
import re
import subprocess
import sys
import tomllib
from pathlib import Path

import msgspec

CONFIG_PATH = Path("accounts.toml")

PROVIDER_PRESETS: dict[str, dict[str, object]] = {
    "gmail": {
        "imap_host": "imap.gmail.com",
        "imap_port": 993,
        "imap_starttls": False,
        "smtp_host": "smtp.gmail.com",
        "smtp_port": 465,
        "drafts_folder": "[Gmail]/Drafts",
    },
    "protonmail-bridge": {
        "imap_host": "127.0.0.1",
        "imap_port": 1143,
        "imap_starttls": True,
        "smtp_host": "127.0.0.1",
        "smtp_port": 1025,
        "drafts_folder": "Drafts",
    },
}


_NON_ACCOUNT_KEYS = frozenset({"watch", "owner"})


class OwnerConfig(msgspec.Struct):
    github_user: str
    name: str = ""


class WatchConfig(msgspec.Struct):
    poll_interval: int = 300
    notify: bool = False


class Account(msgspec.Struct):
    provider: str = "imap"
    user: str = ""
    password: str = ""
    password_cmd: str = ""
    labels: list[str] = []
    imap_host: str = ""
    imap_port: int = 993
    imap_starttls: bool = False
    smtp_host: str = ""
    smtp_port: int = 465
    drafts_folder: str = "Drafts"
    sync_days: int = 3650
    default: bool = False


def _apply_preset(account: Account) -> Account:
    """Merge provider preset defaults — account values win over preset."""
    preset = PROVIDER_PRESETS.get(account.provider)
    if preset is None:
        return account
    # Only apply preset values for fields still at their Struct defaults
    defaults = Account()
    updates: dict[str, object] = {}
    for key, preset_val in preset.items():
        if getattr(account, key) == getattr(defaults, key):
            updates[key] = preset_val
    if updates:
        return msgspec.structs.replace(account, **updates)
    return account


def resolve_password(account: Account) -> str:
    """Return password: inline value if set, else run password_cmd."""
    if account.password:
        return account.password
    if account.password_cmd:
        result = subprocess.run(
            account.password_cmd,
            shell=True,
            capture_output=True,
            text=True,
            check=True,
        )
        return result.stdout.strip()
    raise ValueError(f"Account {account.user!r} has no password or password_cmd")


def load_accounts(path: Path | None = None) -> dict[str, Account]:
    """Parse accounts.toml → {name: Account} mapping."""
    if path is None:
        path = CONFIG_PATH
    if not path.exists():
        return {}
    with open(path, "rb") as f:
        raw = tomllib.load(f)
    accounts_section = raw.get("accounts", raw)
    result: dict[str, Account] = {}
    for name, data in accounts_section.items():
        if name in _NON_ACCOUNT_KEYS:
            continue
        account = msgspec.convert(data, Account)
        account = _apply_preset(account)
        result[name] = account
    return result


def _legacy_account_from_env() -> Account:
    """Build a synthetic Account from legacy .env GMAIL_* vars."""
    from dotenv import load_dotenv

    load_dotenv()

    user = os.environ.get("GMAIL_USER_EMAIL", "")
    password = os.environ.get("GMAIL_APP_PASSWORD", "")
    labels_str = os.environ.get("GMAIL_SYNC_LABELS", "")
    sync_days = int(os.environ.get("GMAIL_SYNC_DAYS", "3650"))

    if not user:
        raise SystemExit("No accounts.toml found and GMAIL_USER_EMAIL not set in .env")

    return Account(
        provider="gmail",
        user=user,
        password=password.replace(" ", ""),
        labels=[s.strip() for s in labels_str.split(",") if s.strip()],
        imap_host="imap.gmail.com",
        imap_port=993,
        smtp_host="smtp.gmail.com",
        smtp_port=465,
        drafts_folder="[Gmail]/Drafts",
        sync_days=sync_days,
        default=True,
    )


def load_accounts_or_env(path: Path | None = None) -> dict[str, Account]:
    """Load accounts.toml, falling back to .env GMAIL_* vars."""
    accounts = load_accounts(path)
    if accounts:
        return accounts
    return {"_legacy": _legacy_account_from_env()}


def load_owner(path: Path | None = None) -> OwnerConfig:
    """Load [owner] section from accounts.toml."""
    if path is None:
        path = CONFIG_PATH
    if not path.exists():
        raise SystemExit(
            f"accounts.toml not found at {path}.\n"
            "Add an [owner] section with github_user."
        )
    with open(path, "rb") as f:
        raw = tomllib.load(f)
    owner_data = raw.get("owner")
    if owner_data is None:
        raise SystemExit(
            "Missing [owner] section in accounts.toml.\n"
            'Add: [owner]\ngithub_user = "your-github-username"'
        )
    return msgspec.convert(owner_data, OwnerConfig)


def get_default_account(accounts: dict[str, Account]) -> tuple[str, Account]:
    """Return (name, account) for the default account."""
    for name, acct in accounts.items():
        if acct.default:
            return name, acct
    # Fall back to first account
    name = next(iter(accounts))
    return name, accounts[name]


def get_account_for_email(
    accounts: dict[str, Account], email_addr: str
) -> tuple[str, Account] | None:
    """Lookup account by email address."""
    email_lower = email_addr.lower()
    for name, acct in accounts.items():
        if acct.user.lower() == email_lower:
            return name, acct
    return None


def add_label_to_account(
    account_name: str,
    label: str,
    path: Path | None = None,
) -> bool:
    """Add a label to an account's labels list in accounts.toml.

    Does a text-level edit to preserve comments and formatting.
    Returns True if the label was added, False if already present.
    """
    if path is None:
        path = CONFIG_PATH
    if not path.exists():
        print(
            f"accounts.toml not found at {path}",
            file=sys.stderr,
        )
        sys.exit(1)

    # Verify account exists and label isn't already there
    accounts = load_accounts(path)
    if account_name not in accounts:
        print(
            f"Unknown account: {account_name}\nAvailable: {', '.join(accounts.keys())}",
            file=sys.stderr,
        )
        sys.exit(1)
    if label in accounts[account_name].labels:
        return False

    # Text-level edit: find the labels line for this account section
    text = path.read_text(encoding="utf-8")

    # Find the account section header
    section_re = re.compile(
        rf"^\[accounts\.{re.escape(account_name)}\]",
        re.MULTILINE,
    )
    section_match = section_re.search(text)
    if not section_match:
        # Try flat format (no [accounts.] prefix)
        section_re = re.compile(rf"^\[{re.escape(account_name)}\]", re.MULTILINE)
        section_match = section_re.search(text)
    if not section_match:
        print(
            f"Could not find [{account_name}] section in {path}",
            file=sys.stderr,
        )
        sys.exit(1)

    # Find the labels = [...] line after the section header
    labels_re = re.compile(
        r"^(labels\s*=\s*\[)(.*?)(\])",
        re.MULTILINE,
    )
    # Search from the section start
    labels_match = labels_re.search(text, section_match.end())
    if not labels_match:
        print(
            f"Could not find labels line for account {account_name}",
            file=sys.stderr,
        )
        sys.exit(1)

    # Check it's within this section (before next section header)
    next_section = re.search(r"^\[", text[section_match.end() :], re.MULTILINE)
    beyond_section = (
        next_section
        and labels_match.start() > section_match.end() + next_section.start()
    )
    if beyond_section:
        print(
            f"Could not find labels line for account {account_name}",
            file=sys.stderr,
        )
        sys.exit(1)

    # Append the label
    existing = labels_match.group(2).strip()
    if existing:
        new_labels = f'{existing}, "{label}"'
    else:
        new_labels = f'"{label}"'

    new_text = (
        text[: labels_match.start()]
        + f"labels = [{new_labels}]"
        + text[labels_match.end() :]
    )
    path.write_text(new_text, encoding="utf-8")
    return True


def add_label_main() -> None:
    """CLI: corrkit add-label LABEL --account ACCOUNT"""
    parser = argparse.ArgumentParser(
        description="Add a label to an account's sync config"
    )
    parser.add_argument("label", help="Label to add")
    parser.add_argument(
        "--account",
        required=True,
        help="Account name in accounts.toml",
    )
    args = parser.parse_args()

    added = add_label_to_account(args.account, args.label)
    if added:
        print(f"Added '{args.label}' to account '{args.account}' in accounts.toml")
    else:
        print(f"Label '{args.label}' already in account '{args.account}'")


def load_watch_config(path: Path | None = None) -> WatchConfig:
    """Load [watch] section from accounts.toml. Returns defaults if missing."""
    if path is None:
        path = CONFIG_PATH
    if not path.exists():
        return WatchConfig()
    with open(path, "rb") as f:
        raw = tomllib.load(f)
    watch_data = raw.get("watch")
    if watch_data is None:
        return WatchConfig()
    return msgspec.convert(watch_data, WatchConfig)

File: src/accounts.rs
//! Account configuration — parse accounts.toml with provider presets.

use anyhow::{bail, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};

use crate::resolve;

/// Provider presets for common IMAP/SMTP configurations.
pub fn provider_presets() -> HashMap<&'static str, AccountDefaults> {
    let mut m = HashMap::new();
    m.insert(
        "gmail",
        AccountDefaults {
            imap_host: "imap.gmail.com",
            imap_port: 993,
            imap_starttls: false,
            smtp_host: "smtp.gmail.com",
            smtp_port: 465,
            drafts_folder: "[Gmail]/Drafts",
        },
    );
    m.insert(
        "protonmail-bridge",
        AccountDefaults {
            imap_host: "127.0.0.1",
            imap_port: 1143,
            imap_starttls: true,
            smtp_host: "127.0.0.1",
            smtp_port: 1025,
            drafts_folder: "Drafts",
        },
    );
    m
}

pub struct AccountDefaults {
    pub imap_host: &'static str,
    pub imap_port: u16,
    pub imap_starttls: bool,
    pub smtp_host: &'static str,
    pub smtp_port: u16,
    pub drafts_folder: &'static str,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OwnerConfig {
    pub github_user: String,
    #[serde(default)]
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WatchConfig {
    #[serde(default = "default_poll_interval")]
    pub poll_interval: u64,
    #[serde(default)]
    pub notify: bool,
}

fn default_poll_interval() -> u64 {
    300
}

impl Default for WatchConfig {
    fn default() -> Self {
        Self {
            poll_interval: 300,
            notify: false,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Account {
    #[serde(default = "default_provider")]
    pub provider: String,
    #[serde(default)]
    pub user: String,
    #[serde(default)]
    pub password: String,
    #[serde(default)]
    pub password_cmd: String,
    #[serde(default)]
    pub labels: Vec<String>,
    #[serde(default)]
    pub imap_host: String,
    #[serde(default = "default_imap_port")]
    pub imap_port: u16,
    #[serde(default)]
    pub imap_starttls: bool,
    #[serde(default)]
    pub smtp_host: String,
    #[serde(default = "default_smtp_port")]
    pub smtp_port: u16,
    #[serde(default = "default_drafts_folder")]
    pub drafts_folder: String,
    #[serde(default = "default_sync_days")]
    pub sync_days: u32,
    #[serde(default)]
    pub default: bool,
}

fn default_provider() -> String {
    "imap".to_string()
}
fn default_imap_port() -> u16 {
    993
}
fn default_smtp_port() -> u16 {
    465
}
fn default_drafts_folder() -> String {
    "Drafts".to_string()
}
fn default_sync_days() -> u32 {
    3650
}

impl Default for Account {
    fn default() -> Self {
        Self {
            provider: "imap".to_string(),
            user: String::new(),
            password: String::new(),
            password_cmd: String::new(),
            labels: vec![],
            imap_host: String::new(),
            imap_port: 993,
            imap_starttls: false,
            smtp_host: String::new(),
            smtp_port: 465,
            drafts_folder: "Drafts".to_string(),
            sync_days: 3650,
            default: false,
        }
    }
}

/// Apply provider preset defaults. Account values win over preset.
fn apply_preset(account: &mut Account) {
    let presets = provider_presets();
    let Some(preset) = presets.get(account.provider.as_str()) else {
        return;
    };
    let defaults = Account::default();
    if account.imap_host == defaults.imap_host {
        account.imap_host = preset.imap_host.to_string();
    }
    if account.imap_port == defaults.imap_port {
        account.imap_port = preset.imap_port;
    }
    if account.imap_starttls == defaults.imap_starttls && preset.imap_starttls {
        account.imap_starttls = preset.imap_starttls;
    }
    if account.smtp_host == defaults.smtp_host {
        account.smtp_host = preset.smtp_host.to_string();
    }
    if account.smtp_port == defaults.smtp_port {
        account.smtp_port = preset.smtp_port;
    }
    if account.drafts_folder == defaults.drafts_folder {
        account.drafts_folder = preset.drafts_folder.to_string();
    }
}

/// Resolve password: inline value if set, else run password_cmd.
pub fn resolve_password(account: &Account) -> Result<String> {
    if !account.password.is_empty() {
        return Ok(account.password.clone());
    }
    if !account.password_cmd.is_empty() {
        let output = std::process::Command::new("sh")
            .arg("-c")
            .arg(&account.password_cmd)
            .output()?;
        if !output.status.success() {
            bail!(
                "password_cmd failed: {}",
                String::from_utf8_lossy(&output.stderr).trim()
            );
        }
        return Ok(String::from_utf8_lossy(&output.stdout).trim().to_string());
    }
    bail!("Account {:?} has no password or password_cmd", account.user)
}

const NON_ACCOUNT_KEYS: &[&str] = &["watch", "owner", "routing", "mailboxes"];

/// Parse accounts from .corky.toml or accounts.toml → {name: Account} mapping.
///
/// Resolution order (when path is None):
/// 1. .corky.toml / corky.toml (new unified config)
/// 2. accounts.toml (legacy)
pub fn load_accounts(path: Option<&Path>) -> Result<HashMap<String, Account>> {
    let path = match path {
        Some(p) => PathBuf::from(p),
        None => {
            let ck = resolve::corky_toml();
            if ck.exists() {
                ck
            } else {
                resolve::accounts_toml()
            }
        }
    };
    if !path.exists() {
        return Ok(HashMap::new());
    }
    let content = std::fs::read_to_string(&path)?;
    let raw: toml::Value = toml::from_str(&content)?;
    let table = raw.as_table().unwrap();

    // Try [accounts.*] format first, fall back to flat format
    let accounts_section = if let Some(toml::Value::Table(accts)) = table.get("accounts") {
        accts.clone()
    } else {
        table.clone()
    };

    let mut result = HashMap::new();
    for (name, data) in &accounts_section {
        if NON_ACCOUNT_KEYS.contains(&name.as_str()) {
            continue;
        }
        if !data.is_table() {
            continue;
        }
        let mut account: Account = data.clone().try_into()?;
        apply_preset(&mut account);
        result.insert(name.clone(), account);
    }
    Ok(result)
}

/// Build a synthetic Account from legacy .env GMAIL_* vars.
fn legacy_account_from_env() -> Result<Account> {
    dotenvy::dotenv().ok();
    let user = std::env::var("GMAIL_USER_EMAIL")
        .map_err(|_| anyhow::anyhow!("No accounts.toml found and GMAIL_USER_EMAIL not set in .env"))?;
    let password = std::env::var("GMAIL_APP_PASSWORD")
        .unwrap_or_default()
        .replace(' ', "");
    let labels_str = std::env::var("GMAIL_SYNC_LABELS").unwrap_or_default();
    let sync_days: u32 = std::env::var("GMAIL_SYNC_DAYS")
        .unwrap_or_else(|_| "3650".to_string())
        .parse()
        .unwrap_or(3650);

    Ok(Account {
        provider: "gmail".to_string(),
        user,
        password,
        labels: labels_str
            .split(',')
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .collect(),
        imap_host: "imap.gmail.com".to_string(),
        imap_port: 993,
        smtp_host: "smtp.gmail.com".to_string(),
        smtp_port: 465,
        drafts_folder: "[Gmail]/Drafts".to_string(),
        sync_days,
        default: true,
        ..Default::default()
    })
}

/// Load accounts.toml, falling back to .env GMAIL_* vars.
pub fn load_accounts_or_env(path: Option<&Path>) -> Result<HashMap<String, Account>> {
    let accounts = load_accounts(path)?;
    if !accounts.is_empty() {
        return Ok(accounts);
    }
    let legacy = legacy_account_from_env()?;
    let mut m = HashMap::new();
    m.insert("_legacy".to_string(), legacy);
    Ok(m)
}

/// Load [owner] section from .corky.toml or accounts.toml.
pub fn load_owner(path: Option<&Path>) -> Result<OwnerConfig> {
    let path = match path {
        Some(p) => PathBuf::from(p),
        None => {
            let ck = resolve::corky_toml();
            if ck.exists() {
                ck
            } else {
                resolve::accounts_toml()
            }
        }
    };
    if !path.exists() {
        bail!(
            "Config not found at {}.\nRun 'corky init' or add an [owner] section with github_user.",
            path.display()
        );
    }
    let content = std::fs::read_to_string(&path)?;
    let raw: toml::Value = toml::from_str(&content)?;
    let owner_data = raw
        .get("owner")
        .ok_or_else(|| {
            anyhow::anyhow!(
                "Missing [owner] section in config.\nAdd: [owner]\ngithub_user = \"your-github-username\""
            )
        })?;
    let owner: OwnerConfig = owner_data.clone().try_into()?;
    Ok(owner)
}

/// Return (name, account) for the default account.
pub fn get_default_account(accounts: &HashMap<String, Account>) -> Result<(String, Account)> {
    for (name, acct) in accounts {
        if acct.default {
            return Ok((name.clone(), acct.clone()));
        }
    }
    // Fall back to first account
    let (name, acct) = accounts.iter().next().ok_or_else(|| anyhow::anyhow!("No accounts configured"))?;
    Ok((name.clone(), acct.clone()))
}

/// Lookup account by email address.
pub fn get_account_for_email(
    accounts: &HashMap<String, Account>,
    email_addr: &str,
) -> Option<(String, Account)> {
    let email_lower = email_addr.to_lowercase();
    for (name, acct) in accounts {
        if acct.user.to_lowercase() == email_lower {
            return Some((name.clone(), acct.clone()));
        }
    }
    None
}

/// Add a label to an account's labels list in .corky.toml or accounts.toml.
///
/// Uses toml_edit for format-preserving edits.
/// Returns Ok(true) if added, Ok(false) if already present.
pub fn add_label_to_account(account_name: &str, label: &str, path: Option<&Path>) -> Result<bool> {
    let path = match path {
        Some(p) => PathBuf::from(p),
        None => {
            let ck = resolve::corky_toml();
            if ck.exists() {
                ck
            } else {
                resolve::accounts_toml()
            }
        }
    };
    if !path.exists() {
        bail!("Config not found at {}", path.display());
    }

    // Verify account exists and label isn't already there
    let accounts = load_accounts(Some(&path))?;
    let Some(acct) = accounts.get(account_name) else {
        bail!(
            "Unknown account: {}\nAvailable: {}",
            account_name,
            accounts.keys().cloned().collect::<Vec<_>>().join(", ")
        );
    };
    if acct.labels.contains(&label.to_string()) {
        return Ok(false);
    }

    // Format-preserving edit with toml_edit
    let content = std::fs::read_to_string(&path)?;
    let mut doc = content.parse::<toml_edit::DocumentMut>()?;

    // Try [accounts.{name}] first, then [{name}]
    let labels_array = if let Some(accounts_table) = doc.get_mut("accounts") {
        accounts_table
            .get_mut(account_name)
            .and_then(|t| t.get_mut("labels"))
    } else {
        doc.get_mut(account_name)
            .and_then(|t| t.get_mut("labels"))
    };

    if let Some(labels) = labels_array {
        if let Some(arr) = labels.as_array_mut() {
            arr.push(label);
        }
    }

    std::fs::write(&path, doc.to_string())?;
    Ok(true)
}

/// CLI: corky add-label LABEL --account ACCOUNT
pub fn add_label_cmd(label: &str, account: &str) -> Result<()> {
    let added = add_label_to_account(account, label, None)?;
    if added {
        println!("Added '{}' to account '{}'", label, account);
    } else {
        println!("Label '{}' already in account '{}'", label, account);
    }
    Ok(())
}

/// Load [watch] section from .corky.toml or accounts.toml. Returns defaults if missing.
pub fn load_watch_config(path: Option<&Path>) -> Result<WatchConfig> {
    let path = match path {
        Some(p) => PathBuf::from(p),
        None => {
            let ck = resolve::corky_toml();
            if ck.exists() {
                ck
            } else {
                resolve::accounts_toml()
            }
        }
    };
    if !path.exists() {
        return Ok(WatchConfig::default());
    }
    let content = std::fs::read_to_string(&path)?;
    let raw: toml::Value = toml::from_str(&content)?;
    match raw.get("watch") {
        Some(watch_data) => {
            let config: WatchConfig = watch_data.clone().try_into()?;
            Ok(config)
        }
        None => Ok(WatchConfig::default()),
    }
}
